### PR TITLE
Rewrite and overhaul of AmplNLWriter.

### LINKED ISSUE #15: Commented out functions in nl_params.jl
What's the reasoning behind commenting out some of the functions in `nl_params.jl`? (i.e. [min()](https://github.com/JackDunnNZ/AmplNLWriter.jl/blob/master/src/nl_params.jl#L9))

### LINKED ISSUE #72: clean_solverdata and tmp files
At some point this package generates `tmp*` files in the `.solverdata` directory.  I did not see the proper way to clean these.  `clean_solverdata` focuses on `.nl` and `.sol` files.

### LINKED ISSUE #108: Julia Scratch Spaces
Julia v1.6 will introduce a new scratch space feature for packages.  Possibly this could be useful to this package as the place where the `nl` files are stored for execution.

### PR DESCRIPTION
**Pull request by odow**
Ideally, some, or all of this would be migrated into
MOI.FileFormats as an NL submodule.

<s>This needs many more tests before merging.</s>

This implementation is simplier than the current one, because it
doesn't try to detect or simplify linear expressions.

Closes #15 (They're commented out for a reason, and now much better explained)
Closes #72 
Closes #108 (We don't need persistent storage between sessions.)
Closes #123

### CODE REVIEW

**Comment by odow** [comment]
Okay, this is now passing all tests so I'm pretty confident there aren't any lingering bugs. 

I still need to benchmark against the old version, but I'd expect this version to be significantly quicker. I have some ideas for other improvements, but that can wait for a different PR.

On the whole, this version is much more maintainable. A lot of the core code was untouched since 2016, and I've added a lot of comments throughout on the format. Unfortunately the PDFs aren't the most descriptive for some features.

File: src/AmplNLWriter.jl [hunk 1]
```diff
@@ -1,934 +1,266 @@
...
+    empty!(model.single_variable_mask)
+    empty!(model.lower_bound)
+    empty!(model.upper_bound)
+    empty!(model.var_to_name)
+    model.name_to_var = nothing
+    model.nextconstraintid = 0
+    empty!(model.con_to_name)
+    model.name_to_con = nothing
+    empty!(model.constrmap)
+    MOI.empty!(model.moi_scalaraffinefunction)
```
Comment (blegat) [inline src/AmplNLWriter.jl:114]: My using `ext` and not `UniversalFallback` ?

Comment (odow) [reply src/AmplNLWriter.jl:114]: We would still have the issue with `MOI.empty!`. This is a nicer solution: https://github.com/jump-dev/MathOptInterface.jl/pull/1295

Comment (blegat) [reply src/AmplNLWriter.jl:115]: Why would we still have the issue? If we start storing part of the model in `ext` then all getters are wrong and we should have `get_without_ext` too

Comment (odow) [reply src/AmplNLWriter.jl:115]: I implemented the getters.
https://github.com/jump-dev/AmplNLWriter.jl/blob/e1bdc69b89675744ed93f784cf5d8b79d9b1608b/src/AmplNLWriter.jl#L137-L148

Basically, I want a `Model` that I can store extra stuff in, like the `amplexe` function, and I want to be able to, on `MOI.empty!`, modify some of that extra stuff without deleting it.

Comment (odow) [reply src/AmplNLWriter.jl:115]: I guess I could use the current `_NLModel` as `Optimizer`, and only implement `copy_to`.

**Comment by odow** [comment]
Okay, the design is a lot better now. I don't know why I didn't consider it before. This PR is large and unwieldy, so I'm going to merge it as is, and make some smaller follow-up PRs.

**Pull request merged**